File: Solution/Python/The-Magical-Forest-of-Algora.py
class DanceMove:
    Twirl = 0
    Leap = 1
    Spin = 2
    Slide = 3

class Creature:
    def __init__(self, name, dance_moves):
        self.name = name
        self.dance_moves = dance_moves

dance_moves = {
    DanceMove.Twirl: "Twirl",
    DanceMove.Leap: "Leap",
    DanceMove.Spin: "Spin",
    DanceMove.Slide: "Slide",
}

effects = {
    '0,0,0': "A flurry of autumn leaves fills the air.",
    '0,0,1': "A soft mist envelops the forest.",
    '0,0,2': "A chorus of crickets begins to play.",
    '0,0,3': "Fireflies light up the forest.",
    '0,1,0': "A gentle breeze rustles the leaves.",
    '0,1,1': "A rainbow forms in the morning dew.",
    '0,1,2': "A waterfall appears from nowhere.",
    '0,1,3': "A flock of birds takes flight.",
    '0,2,0': "The forest floor blooms with wildflowers.",
    '0,2,1': "A gentle rain starts falling.",
    '0,2,2': "A swarm of butterflies fills the air.",
    '0,2,3': "A stream begins to babble in harmony with the music.",
    '0,3,0': "The air fills with the scent of fresh flowers.",
    '0,3,1': "A brilliant light illuminates the forest.",
    '0,3,2': "The forest glows with a soft light.",
    '0,3,3': "A chorus of forest creatures begins to sing.",
    '1,0,0': "A soft glow envelops the forest.",
    '1,0,1': "A symphony of bird songs fills the air.",
    '1,0,2': "A cascade of petals falls from the trees.",
    '1,0,3': "A gentle snow begins to fall.",
    '1,1,0': "A rainbow appears in the sky.",
    '1,1,1': "A waterfall springs from the ground.",
    '1,1,2': "A gust of wind blows through the forest.",
    '1,1,3': "A chorus of frogs begins to sing.",
    '1,2,0': "A swarm of bees buzzes in harmony with the music.",
    '1,2,1': "A field of flowers blooms instantly.",
    '1,2,2': "A group of squirrels performs a synchronized dance.",
    '1,2,3': "A group of fireflies begins a light show.",
    '1,3,0': "A group of mushrooms glow brightly.",
    '1,3,1': "A group of trees sway in time with the music.",
    '1,3,2': "A group of rocks levitate in the air.",
    '1,3,3': "A group of clouds forms a beautiful pattern in the sky.",
    '2,0,0': "A constellation of stars appears in the sky.",
    '2,0,1': "A sudden gust of wind lifts leaves into the air.",
    '2,0,2': "A group of rabbits perform a synchronized hop.",
    '2,0,3': "A sudden bloom of flowers covers the forest floor.",
    '2,1,0': "A group of fish leap out of a stream in a beautiful arc.",
    '2,1,1': "A group of trees bend and sway to the music.",
    '2,1,2': "A group of clouds change shape to form a beautiful scene.",
    '2,1,3': "A group of fireflies spell out a message in the air.",
    '2,2,0': "A group of birds fly in a mesmerizing pattern.",
    '2,2,1': "A group of butterflies flutter in a beautiful dance.",
    '2,2,2': "A group of deer prance in a circle.",
    '2,2,3': "A group of frogs leap in a synchronized dance.",
    '2,3,0': "A group of mushrooms suddenly grow and emit a soft glow.",
    '2,3,1': "A group of trees suddenly grow fruit.",
    '2,3,2': "A group of flowers bloom and emit a beautiful fragrance.",
    '2,3,3': "A group of rocks suddenly sparkle with gemstones.",
    '3,0,0': "A group of trees suddenly grow taller.",
    '3,0,1': "A group of flowers suddenly change color.",
    '3,0,2': "A group of rocks suddenly become transparent.",
    '3,0,3': "A group of clouds suddenly take on a beautiful color.",
    '3,1,0': "A group of birds suddenly start singing in harmony.",
    '3,1,1': "A group of butterflies suddenly start flying in a beautiful pattern.",
    '3,1,2': "A group of deer suddenly start prancing in a circle.",
    '3,1,3': "A group of frogs suddenly start leaping in a synchronized dance.",
    '3,2,0': "A group of mushrooms suddenly start glowing in different colors.",
    '3,2,1': "A group of trees suddenly start swaying in time with the music.",
    '3,2,2': "A group of flowers suddenly start blooming and emitting a beautiful fragrance.",
    '3,2,3': "A group of rocks suddenly start sparkling with different colored gemstones.",
    '3,3,0': "A group of clouds suddenly start forming different shapes.",
    '3,3,1': "A group of birds suddenly start flying in a mesmerizing pattern.",
    '3,3,2': "A group of butterflies suddenly start fluttering in a beautiful dance.",
    '3,3,3': "A group of frogs suddenly start leaping in a synchronized dance.",
  }

def print_move(name, move):
    move_name = dance_moves.get(move, "Unknown")
    print(f'{name} performs a {move_name},')

lox = Creature("Lox", [DanceMove.Twirl, DanceMove.Leap, DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap])
drako = Creature("Drako", [DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap, DanceMove.Leap, DanceMove.Spin])
flix = Creature("Flix", [DanceMove.Slide, DanceMove.Spin, DanceMove.Twirl, DanceMove.Slide, DanceMove.Leap])

for i in range(5):
    lox_move = lox.dance_moves[i]
    drako_move = drako.dance_moves[i]
    flix_move = flix.dance_moves[i]

    print_move(lox.name, lox_move)
    print_move(drako.name, drako_move)
    print_move(flix.name, flix_move)

    print()
    effect = effects.get(f'{lox_move},{drako_move},{flix_move}', "A magical effect happens, but it's hard to describe.")
    print(f'✨ {effect} ✨')
    print()

def print_move(name, move):
    move_name = dance_moves.get(move, "Unknown")
    print(f'{name} performs a {move_name},')

File: Solution/CSharp/The-Magical-Forest-of-Algora.cs
using System;
using System.Collections.Generic;

enum DanceMove
{
    Twirl,
    Leap,
    Spin,
    Slide
}

class Creature
{
    public required string Name { get; set; }
    public required List<DanceMove> DanceMoves { get; set; }
}

class Program
{
    private static Dictionary<(DanceMove, DanceMove, DanceMove), string> _effects = new()
    {
        [(DanceMove.Twirl, DanceMove.Twirl, DanceMove.Twirl)] = "A flurry of autumn leaves fills the air.",
        [(DanceMove.Twirl, DanceMove.Twirl, DanceMove.Leap)] = "A soft mist envelops the forest.",
        [(DanceMove.Twirl, DanceMove.Twirl, DanceMove.Spin)] = "A chorus of crickets begins to play.",
        [(DanceMove.Twirl, DanceMove.Twirl, DanceMove.Slide)] = "Fireflies light up the forest.",
        [(DanceMove.Twirl, DanceMove.Leap, DanceMove.Twirl)] = "A gentle breeze rustles the leaves.",
        [(DanceMove.Twirl, DanceMove.Leap, DanceMove.Leap)] = "A rainbow forms in the morning dew.",
        [(DanceMove.Twirl, DanceMove.Leap, DanceMove.Spin)] = "A waterfall appears from nowhere.",
        [(DanceMove.Twirl, DanceMove.Leap, DanceMove.Slide)] = "A flock of birds takes flight.",
        [(DanceMove.Twirl, DanceMove.Spin, DanceMove.Twirl)] = "The forest floor blooms with wildflowers.",
        [(DanceMove.Twirl, DanceMove.Spin, DanceMove.Leap)] = "A gentle rain starts falling.",
        [(DanceMove.Twirl, DanceMove.Spin, DanceMove.Spin)] = "A swarm of butterflies fills the air.",
        [(DanceMove.Twirl, DanceMove.Spin, DanceMove.Slide)] = "A stream begins to babble in harmony with the music.",
        [(DanceMove.Twirl, DanceMove.Slide, DanceMove.Twirl)] = "The air fills with the scent of fresh flowers.",
        [(DanceMove.Twirl, DanceMove.Slide, DanceMove.Leap)] = "A brilliant light illuminates the forest.",
        [(DanceMove.Twirl, DanceMove.Slide, DanceMove.Spin)] = "The forest glows with a soft light.",
        [(DanceMove.Twirl, DanceMove.Slide, DanceMove.Slide)] = "A chorus of forest creatures begins to sing.",
        [(DanceMove.Leap, DanceMove.Twirl, DanceMove.Twirl)] = "A soft glow envelops the forest.",
        [(DanceMove.Leap, DanceMove.Twirl, DanceMove.Leap)] = "A symphony of bird songs fills the air.",
        [(DanceMove.Leap, DanceMove.Twirl, DanceMove.Spin)] = "A cascade of petals falls from the trees.",
        [(DanceMove.Leap, DanceMove.Twirl, DanceMove.Slide)] = "A gentle snow begins to fall.",
        [(DanceMove.Leap, DanceMove.Leap, DanceMove.Twirl)] = "A rainbow appears in the sky.",
        [(DanceMove.Leap, DanceMove.Leap, DanceMove.Leap)] = "A waterfall springs from the ground.",
        [(DanceMove.Leap, DanceMove.Leap, DanceMove.Spin)] = "A gust of wind blows through the forest.",
        [(DanceMove.Leap, DanceMove.Leap, DanceMove.Slide)] = "A chorus of frogs begins to sing.",
        [(DanceMove.Leap, DanceMove.Spin, DanceMove.Twirl)] = "A swarm of bees buzzes in harmony with the music.",
        [(DanceMove.Leap, DanceMove.Spin, DanceMove.Leap)] = "A field of flowers blooms instantly.",
        [(DanceMove.Leap, DanceMove.Spin, DanceMove.Spin)] = "A group of squirrels performs a synchronized dance.",
        [(DanceMove.Leap, DanceMove.Spin, DanceMove.Slide)] = "A group of fireflies begins a light show.",
        [(DanceMove.Leap, DanceMove.Slide, DanceMove.Twirl)] = "A group of mushrooms glow brightly.",
        [(DanceMove.Leap, DanceMove.Slide, DanceMove.Leap)] = "A group of trees sway in time with the music.",
        [(DanceMove.Leap, DanceMove.Slide, DanceMove.Spin)] = "A group of rocks levitate in the air.",
        [(DanceMove.Leap, DanceMove.Slide, DanceMove.Slide)] = "A group of clouds forms a beautiful pattern in the sky.",
        [(DanceMove.Spin, DanceMove.Twirl, DanceMove.Twirl)] = "A constellation of stars appears in the sky.",
        [(DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap)] = "A sudden gust of wind lifts leaves into the air.",
        [(DanceMove.Spin, DanceMove.Twirl, DanceMove.Spin)] = "A group of rabbits perform a synchronized hop.",
        [(DanceMove.Spin, DanceMove.Twirl, DanceMove.Slide)] = "A sudden bloom of flowers covers the forest floor.",
        [(DanceMove.Spin, DanceMove.Leap, DanceMove.Twirl)] = "A group of fish leap out of a stream in a beautiful arc.",
        [(DanceMove.Spin, DanceMove.Leap, DanceMove.Leap)] = "A group of trees bend and sway to the music.",
        [(DanceMove.Spin, DanceMove.Leap, DanceMove.Spin)] = "A group of clouds change shape to form a beautiful scene.",
        [(DanceMove.Spin, DanceMove.Leap, DanceMove.Slide)] = "A group of fireflies spell out a message in the air.",
        [(DanceMove.Spin, DanceMove.Spin, DanceMove.Twirl)] = "A group of birds fly in a mesmerizing pattern.",
        [(DanceMove.Spin, DanceMove.Spin, DanceMove.Leap)] = "A group of butterflies flutter in a beautiful dance.",
        [(DanceMove.Spin, DanceMove.Spin, DanceMove.Spin)] = "A group of deer prance in a circle.",
        [(DanceMove.Spin, DanceMove.Spin, DanceMove.Slide)] = "A group of frogs leap in a synchronized dance.",
        [(DanceMove.Spin, DanceMove.Slide, DanceMove.Twirl)] = "A group of mushrooms suddenly grow and emit a soft glow.",
        [(DanceMove.Spin, DanceMove.Slide, DanceMove.Leap)] = "A group of trees suddenly grow fruit.",
        [(DanceMove.Spin, DanceMove.Slide, DanceMove.Spin)] = "A group of flowers bloom and emit a beautiful fragrance.",
        [(DanceMove.Spin, DanceMove.Slide, DanceMove.Slide)] = "A group of rocks suddenly sparkle with gemstones.",
    };

    static void Main(string[] args)
    {
        var lox = new Creature
        {
            Name = "Lox",
            DanceMoves = [DanceMove.Twirl, DanceMove.Leap, DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap]
        };

        var drako = new Creature
        {
            Name = "Drako",
            DanceMoves = [DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap, DanceMove.Leap, DanceMove.Spin]
        };

        var flix = new Creature
        {
            Name = "Flix",
            DanceMoves = [DanceMove.Slide, DanceMove.Spin, DanceMove.Twirl, DanceMove.Slide, DanceMove.Leap]
        };

        for (int i = 0; i < 5; i++)
        {
            var loxMove = lox.DanceMoves[i];
            var drakoMove = drako.DanceMoves[i];
            var flixMove = flix.DanceMoves[i];

            PrintMove(lox.Name, loxMove);
            PrintMove(drako.Name, drakoMove);
            PrintMove(flix.Name, flixMove);

            Console.WriteLine("");
            Console.WriteLine(_effects.TryGetValue((loxMove, drakoMove, flixMove), out var effect) ?
                $"✨ {effect} ✨":
                "A magical effect happens, but it's hard to describe.");
            Console.WriteLine();
        }
    }

    static void PrintMove(string name, DanceMove move)
    {
        Console.ForegroundColor = ConsoleColor.Red;
        Console.Write($"{name}");
        Console.ResetColor();
        Console.Write(" performs a ");
        Console.ForegroundColor = ConsoleColor.Yellow;
        Console.Write($"{move}");
        Console.ResetColor();
        Console.Write(", ");
    }
}